src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -883,5 +897,42 @@ private bool IsFailOverable(RequestFailedException rfe)
                    rfe.Status == (int)HttpStatusCode.RequestTimeout ||
                    rfe.Status >= (int)HttpStatusCode.InternalServerError;
         }
+
+        private async Task<Dictionary<string, ConfigurationSetting>> MapConfigurationData(Dictionary<string, ConfigurationSetting> data)
+        {
+            Dictionary<string, ConfigurationSetting> mappedData = new Dictionary<string, ConfigurationSetting>(StringComparer.OrdinalIgnoreCase);
+
+            foreach (KeyValuePair<string, ConfigurationSetting> kvp in data)
+            {
+                ConfigurationSetting setting = kvp.Value;
+                foreach (Func<ConfigurationSetting, ValueTask<ConfigurationSetting>> func in _options.UserDefinedMappers)
+                {
+                    setting = await func(setting).ConfigureAwait(false);
+                }
+                mappedData[kvp.Key] = setting;
+            }
+
+            return mappedData;
+        }
+
+        private void UpdateWatchedSettings()
+        {
+            _watchedSettings.Clear();
+
+            foreach (KeyValueWatcher changeWatcher in _options.ChangeWatchers)
+            {
+                string watchedKey = changeWatcher.Key;
+                string watchedLabel = changeWatcher.Label;
+                KeyValueIdentifier watchedKeyLabel = new KeyValueIdentifier(watchedKey, watchedLabel);
+
+                ConfigurationSetting watchedKvCopy = _serverData.Values.FirstOrDefault((kv) => kv.Key.Equals(watchedKey) && kv.Label == watchedLabel);
+
+                // If the key-value was found, store it for updating the settings
+                if (watchedKvCopy != null)

```

Comment:
>if (watchedKvCopy != null) [](http://example.com/codeflow?start=16&length=26)

The reason watchedKvCopy could be null is when we have multiple refresh registered key-values with same key but different label. When we were populating _watchedSettings in `LoadKeyValuesRegisteredForRefresh`, we would add all refresh registered key-values to _watchedSettings and the overwritten key-values were not lost. That's what we want to achieve here too. We need to rethink how to extract the logic to update _watchedSettings from LoadKeyValuesRegisteredForRefresh, without losing the overwritten key-values.